xen/drivers/passthrough/arm/scmi_smc_maker.c

```diff
@@ -0,0 +1,350 @@
+/*
+ * scmi_process.c
+ *
+ *  Created on: Sep 3, 2024
+ *      Author: amoiseev
+ */
+
+#include <asm/p2m.h>
+#include <xen/device_tree.h>
+#include <xen/init.h>
+#include <xen/iocap.h>
+#include <xen/lib.h>
+#include <xen/libfdt/libfdt.h>
+#include <xen/scmi_smc_maker.h>
+
+#define SCMI_NODE_PATH_MAX_LEN 128
+#define SCMI_NODE_COMPATIBLE   "arm,scmi-smc"
+#define SCMI_SHMEM_COMPATIBLE  "arm,scmi-shmem"
+
+struct scmi_phandle {
+    struct list_head list;
+    uint32_t phandle;
+    char full_name[SCMI_NODE_PATH_MAX_LEN];
+};
+
+LIST_HEAD(scmi_ph_list);
+
+static int __init mem_permit_access(struct domain *d, uint64_t addr, uint64_t len)
+{
+    int res;
+    res = iomem_permit_access(d, paddr_to_pfn(addr),
+            paddr_to_pfn(PAGE_ALIGN(addr + len - 1)));
+    if ( res )
+        return res;
+
+    return map_regions_p2mt(d, gaddr_to_gfn(addr), PFN_DOWN(len),
+            maddr_to_mfn(addr), p2m_mmio_direct_nc);
+}
+
+ int __init make_scmi_shmem_node(struct kernel_info *kinfo)
+{
+    int res;
+    void *fdt = kinfo->fdt;
+    char buf[64];
+    __be32 reg[GUEST_ROOT_ADDRESS_CELLS + GUEST_ROOT_SIZE_CELLS];
+    __be32 *cells;
+    struct domain *d = kinfo->d;
+
+#ifdef CONFIG_ARM_32
+    snprintf(buf, sizeof(buf), "scmi-shmem@%lx",
+            d->arch.sci->paddr);
+#else
+    snprintf(buf, sizeof(buf), "scmi-shmem@%lx",
+            d->arch.sci->paddr);
+#endif
+
+    if ( !is_domain_direct_mapped(d) )
+    {
+        printk(XENLOG_ERR "Non-direct mapped domains doesn't support SCMI\n");
+        return -ENODEV;
+    }
+
+    res = mem_permit_access(kinfo->d, kinfo->d->arch.sci->paddr, 0x1000);
```

Comment:
This should not be part of this file as it seems only DT specific

Reply:
moved out